src/shape_transform_descriptor.cpp

```diff
@@ -0,0 +1,836 @@
+    {
+        // Group axes
+        auto axes_map = group_axes(dimensions);
+        if(axes_map.empty())
+            return;
+
+        renumber_axes(axes_map);
+
+        // Find last axis
+        last_axis = std::prev(axes_map.end())->second.back()->axis;
+
+        // Find missing axes. This will store a mapping between the missing
+        // axis and the next available axis.
+        for(auto axis : range(rank))
+        {
+            if(contains(axes_map, axis))
+                continue;
+            auto it            = axes_map.upper_bound(axis);
+            missing_axes[axis] = it == axes_map.end() ? rank : it->first;
+        }
+    }
+
+    // Find broadcasted dimensions. This will store a map from the next axis
+    // to the previous axes being broadcasted.
+    std::map<std::size_t, std::deque<std::size_t>> broadcast_dims_map;
+    group_find(
+        dimensions.begin(), dimensions.end(), &missing_leading_axis, [&](auto start, auto last) {
+            auto axis = rank;
+            if(last != dimensions.end())
+            {
+                assert(not last->subdimensions.empty());
+                const auto& sub = last->subdimensions.front();
+                assert(not sub.axis.empty());
+                axis = sub.axis.front();
+            }
+            std::deque<std::size_t> dims(std::distance(start, last));
+            std::iota(dims.begin(), dims.end(), std::distance(dimensions.begin(), start));
+            broadcast_dims_map[axis] = dims;
+        });
+
+    // Reinsert removed axis of 1. This tries to insert the missing axis next
+    // to an adjacent axis or used as one of the broadcasted axes in order to
+    // minimize transposition.
+    for(auto&& p : missing_axes)
+    {
+        auto missing_axis = p.first;
+        auto next_axis    = p.second;
+        auto missing_sub  = dimension::sub{1, {missing_axis}};
+        // If next_axis is the rank that means there isnt another axis to
+        // search for, so instead try to insert the axis at the end.
+        if(next_axis == rank)
+        {
+            auto [sub, it, prev] = find_subdimension(
+                *this, [&](const dimension::sub& s) { return s.axis == last_axis; });
+            // Check if we can insert it at the end
+            auto bdims = broadcast_dims_map.find(rank);
+            if(bdims != broadcast_dims_map.end() and not bdims->second.empty())
+            {
+                auto bdim = bdims->second.front();
+                bdims->second.pop_front();
+                set_broadcast_dim(dimensions[bdim], missing_axis);
+            }
+            else
+            {
+                auto next = std::find_if(std::next(it), sub->end(), [&](const dimension::sub& s) {
+                    if(s.len != 1)
+                        return true;
+                    if(s.axis.empty())
+                        return true;
+                    return s.axis.front() > missing_axis;
+                });
```

Comment:
So the `it` is pointer to subdimension with `last_axis`. 
and
```
 // Find last axis
        last_axis = std::prev(axes_map.end())->second.back()->axis;
```

`last_axis` is axis after `renumbering` whereas here `std::find()` is finding it over dimensions without re-numbering. 
How does that work out ?

Reply:
I see `renumbering` is renumbering all subdimensions because it is using pointer to those.